StortSpelProjekt/Game/src/Gamefiles/Components/UpgradeComponents/Upgrades/UpgradeExplosiveAttack.cpp

```diff
@@ -0,0 +1,103 @@
+#include "UpgradeExplosiveAttack.h"
+#include "EngineMath.h"
+#include "ECS/Entity.h"
+#include "Player.h"
+#include "Components/ProjectileComponent.h"
+#include "EnemyFactory.h"
+#include "Misc/AssetLoader.h"
+#include "Renderer/Texture/Texture2DGUI.h"
+#include "ECS/SceneManager.h"
+#include "Renderer/Transform.h"
+#include "ECS/Components/TemporaryLifeComponent.h"
+#include "ECS/Components/TransformComponent.h"
+
+UpgradeExplosiveAttack::UpgradeExplosiveAttack(Entity* parent)
+	:Upgrade(parent)
+{
+	// name the upgrade!
+	SetName("UpgradeExplosiveAttack");
+	// set the upgrade type/types!
+	SetType(F_UpgradeType::RANGE);	// goes on the projectiles
+	// set the price of this upgrade
+	m_Price = 300;
+	m_StartingPrice = m_Price;
+
+	// percentage of damage done to steal as life
+	m_Radius = 3.0;
+	m_PercentageDamage = 0.5;
+
+	m_ImageName = "Explosion.png";
+}
+
+UpgradeExplosiveAttack::~UpgradeExplosiveAttack()
+{
+}
+
+void UpgradeExplosiveAttack::IncreaseLevel()
+{
+	m_Level++;
+	m_Radius = (5 * m_Level);
+	m_PercentageDamage = 0.4 + (0.1 * m_Level);
+	m_Price += m_StartingPrice;
+}
+
+void UpgradeExplosiveAttack::OnRangedHit(Entity* target)
+{
+	//Should only affect enemies. Only enemies should have AiComponent and therefore should be a good test
+	if (target->HasComponent<component::AiComponent>())
```

Comment:
This hasComponent is a very slow function, if there is a better way to check for enemy, use it.